Add App Clip to Demo project

mokagio opened the pull request:
Internal ref pdnsEh-21E-p2

### Description

- Adds an App Clip to the Demo project
- Wires up code signing for the Debug configuration — More work required for the Release build, given it can only be tested on the App Store account and we don't have it set up yet
- Updates the Prototype Build automation to strip the App Clip. The Prototype Build is distributed via the Enterprise account but Enterprise apps are not allowed App Clips.

### Testing Steps

Checkout this branch and run the debug build on device.

Here's a screenshot from mine:

<img alt="IMG_6F3CC09BE35F-1" src="https://github.com/user-attachments/assets/67b5403e-ccbf-4810-ae6f-a5ce2192cfb9" width=200/>

### Where to go from here

For @andrewdmontgomery @pinarol & team to

- Implement the actual App Clip. What's here is just the Xcode template
- Let me / `@Automattic/apps-infrastructure` know whether to set up TestFlight distribution to test the App Clip without having to deploy the app on device. That will require a bunch of Fastlane updates, so I didn't go down that path in the context of this PR.

mokagio committed:
Add App Clip to Demo project

mokagio committed:
Rename App Clip folder to `Demo-App-Clip`

mokagio committed:
Use correct bundle identifier for App Clip target

mokagio committed:
Add `appclips:gravatar.com` associated domain to both app and clip

mokagio committed:
Remove snake casing from App Clip SwiftUI `App`

mokagio committed:
Rename App Clip entitlement to match folder name

mokagio committed:
Configure provisioning profile for Debug – Includes entitlements fix

dangermattic commented:
<!--
  1 Error: You need to add at least one l...
  0 Warnings
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">You need to add at least one label to this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio committed:
Update dummy text in App Clip `ContentView.swift`

This is only to show a screenshot of the App Clip running on my device that
is obviously the one in this codebase.

mokagio committed:
Update automation to strip App Clip when building prototype build

mokagio committed:
Update Fastlane to latest version, 2.226.0

I run into an issue testing the prototype build automation and before
trying to debug it, I thought I'd make sure we're on the latest tooling.

mokagio committed:
Move Demo app entitlement to Demo sources folder for consistency

For consistency with where the App Clip entitlements file is, that is.

mokagio commented on fastlane/Fastfile:
```diff
@@ -213,3 +217,32 @@ def read_from_xcconfig!(key:, xcconfig:)
   configs[key]
 end
+
+# rubocop:disable Metrics/AbcSize
+# rubocop:disable Metrics/CyclomaticComplexity
+def remove_app_clip_dependency!
+  project_path = XCODEPROJ_PATH
+  project = Xcodeproj::Project.open(project_path)
+
+  native_targets = project.targets.select { |t| t.is_a?(Xcodeproj::Project::Object::PBXNativeTarget) }
+  app_targets = native_targets.select { |t| t.product_type == 'com.apple.product-type.application' }
+  app_clip_targets = native_targets.select { |t| t.product_type == 'com.apple.product-type.application.on-demand-install-capable' }
+
+  if app_clip_targets.empty?
+    UI.user_error!("No on-demand-install-capable targets found in #{project_path}. Has something changed in the project structure? Aborting App Clips removal.")
+  end
+
+  app_targets.each do |app|
+    # Delete any dependency of the app target that is an app_clip target
+    app.dependencies.delete_if { |dep| app_clip_targets.include?(dep.target) }
+    # Delete the "Embed App Clips" build phase
+    app.build_phases.delete_if { |b| b.display_name == 'Embed App Clips' }
+  end
+
+  UI.message('Saving the project file....')
+  project.save
+
+  UI.success('Dependency successfuly removed.')
+end
+# rubocop:enable Metrics/AbcSize
+# rubocop:enable Metrics/CyclomaticComplexity
```
This is duplicated from Pocket Casts. We might want to centralize at the release toolkit level, see https://github.com/wordpress-mobile/release-toolkit/issues/629

andrewdmontgomery replied:
Oh right, the original development profile is for an Enterprise app, which doesn't support App Clips.  Thanks for this.

AliSoftware replied:
~~@mokagio given that's the second repo/app for which we need to do that trick (removing AppClip for Enterprise / Prototype builds to compile), worth P2-ing about it for future reference?~~

NVM, I missed your comment above suggesting to centralize this at the `release-toolkit` level, which would indeed be an even better idea 👍

pinarol commented:
Thank you very much! This is very useful for us. I won't merge it for now as this effort might move to a separate repo.

pinarol commented:
@mokagio The app clip target works fine but this time, I can't get the old demo app target to build in this branch. I get the error: `The CFBundleVersion of an App Clip ('1') must match that of its containing parent app ('0').`

pinarol commented:
Also, we need to test OAuth in the app clip target so we'll need this config. Does this require updating the provisioning profile also? @mokagio 

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>com.apple.developer.associated-domains</key>
	<array>
		<string>appclips:gravatar.com</string>
		<string>applinks:gravatar.com</string>
		<string>webcredentials:gravatar.com</string>
	</array>
	<key>com.apple.developer.parent-application-identifiers</key>
	<array>
		<string>$(AppIdentifierPrefix)com.automattic.gravatar-sdk-demo-uikit</string>
	</array>
</dict>
</plist>

```

andrewdmontgomery committed:
Change App Clip CURRENT_PROJECT_VERSION (CFBundleVersion) to match App

andrewdmontgomery commented:
> @mokagio The app clip target works fine but this time, I can't get the old demo app target to build in this branch. I get the error: `The CFBundleVersion of an App Clip ('1') must match that of its containing parent app ('0').`

I fixed this. [[5e5a614](https://github.com/Automattic/Gravatar-SDK-iOS/pull/659/commits/5e5a61439b4e48868b5e8433e1508ed39841d098)]

andrewdmontgomery commented:
> Does this require updating the provisioning profile also?

Ah, yes, I believe it does.  Apologies, I didn't consider that when I opened the request.  Thanks for catching it.

andrewdmontgomery commented:
This happens on a device, when I attempt to install/run the Demo App after having installed/ran the App Clip:

![image](https://github.com/user-attachments/assets/3b787e77-4cad-4b27-a248-8a7360e6f230)

Edit: I had more than one previous version of the "Gravatar Demo" apps installed, from back when we had two demo targets (SwiftUI and UIKit).  Removed.  Problem solved.

mokagio committed:
Move project-level `xcconfig` for demo app to dedicated folder

This will make it clearer that they are shared once we'll read them from
the app clip extension, too.

mokagio committed:
Inherit App Clip version from `Base.xcconfig`

mokagio committed:
Inherit App Clip `IPHONEOS_DEPLOYMENT_TARGET` from project-level setting

mokagio committed:
DRY Demo project `IPHONEOS_DEPLOYMENT_TARGET` to `Base.xcconfig`

mokagio committed:
Inherit App target `CURRENT_PROJECT_VERSION` from `xcconfig`

mokagio committed:
Add associated domains for OAUth to App Clip

mokagio commented:
> Also, we need to test OAuth in the app clip target so we'll need this config. Does this require updating the provisioning profile also? @mokagio

@pinarol @andrewdmontgomery 

51b5a1ac addresses this.

The app clip already had the associated domains capability because of the `appclips:gravatar.com` domain, so no update at the developer portal level was needed.

dangermattic commented:
<!--
  1 Error: This PR is tagged with <code>d...
  0 Warnings
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">This PR is tagged with <code>do not merge</code> label(s).</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

andrewdmontgomery committed:
Leverage `xcconfig` for App Clip project, too (#663)

mokagio commented:
Hey @andrewdmontgomery @pinarol 👋 

Wanted to check in on the state of this. Pinar mentioned "this effort might move to a separate repo," did you make a decision on that?

Not trying to put pressure on you, just touching base. As far as I'm concerned, this PR can remain open indefinitely. No issue on my end.

pinarol commented:
@mokagio This PR has been super useful, but I think we shouldn't merge it. We can keep the branch a while longer, but I think we can close the PR. WDYT @andrewdmontgomery ?

mokagio commented:
Sounds good to me @pinarol
